File: server/src/message.rs
use model::data;

#[derive(Serialize, Deserialize)]
#[derive(Copy, Clone, Debug)]
pub struct Block {
    material: data::Material,
    location: data::Location,
}

#[derive(Message)]
#[derive(Serialize)]
#[derive(Clone, Debug)]
#[serde(tag = "type", deny_unknown_fields)]
pub enum Outgoing {
    ChunkData(data::Mesh),

    BlockData {
        index: data::Index,
        block: Block,
    },

    EntityData {
        id: usize,
        position: data::Position,
    },
}

#[derive(Message)]
#[derive(Deserialize)]
#[derive(Clone, Debug)]
#[serde(tag = "type", deny_unknown_fields)]
pub enum Incoming {
    CreateBlock {
        index: data::Index,   
        block: Block,
    },

    RemoveBlock {
        index: data::Index,   
        block: Block,
    },

    MoveData {
        direction: data::Direction,
    },
}

File: message.js
class Block {
    constructor(material, location) {
        this.material = material;
        this.location = location;
    }
}

class BlockData {
    constructor(index, block) {
        this.index = index;
        this.block = block;
    }
}

// Outgoing (client -> server) data types
class MoveData {
    constructor(direction) {
        this.type = "MoveData";
        this.direction = direction;
    }
}

class CreateBlock {
    constructor(block) {
        this.type = "CreateBlock";
        this.block = block;
    }
}

class RemoveBlock {
    constructor(block) {
        this.type = "RemoveBlock";
        this.block = block;
    }
}
